Lib/_pyio.py

```diff
@@ -1253,6 +1253,17 @@ def seek(self, pos, whence=0):
             self._flush_unlocked()
             return _BufferedIOMixin.seek(self, pos, whence)
 
+    def close(self):
+        with self._write_lock:
+            if self.raw is None or self.closed:
+                return
+        try:
```

Comment:
This doesn't look right. The `close()` method should take the lock once, not take it three times in a row (one of them implicit in `flush()`). Note `_flush_unlocked()` can help you here.

Reply:
That made a bunch of tests (`test_flush_error_on_close`, etc.) fail since `close()` no longer calls `flush()` if the file object is subclassed or `fileobj.flush` is set to something. 

This is actually the same behavior as `buffered_close` in the C implementation. I have added a comment to that effect.